File: deno/examples/a2a/agent_card.ts
// Example: Agent Card Construction
// Demonstrates building A2A AgentCards with capabilities, skills,
// security schemes, and JSON serialization with round-trip verification.
// Run: deno run deno/examples/a2a/agent_card.ts

import type {
  AgentCapabilities,
  AgentCard,
  AgentProvider,
  AgentSkill,
  HttpAuthSecurityScheme,
  SecurityScheme,
} from "@rullama/a2a";

async function main(): Promise<void> {
  // -----------------------------------------------------------------------
  // 1. Build a full AgentCard with all fields populated
  // -----------------------------------------------------------------------
  console.log("=== Full AgentCard ===\n");

  const securitySchemes: Record<string, SecurityScheme> = {
    bearer_auth: {
      httpAuthSecurityScheme: {
        scheme: "Bearer",
        bearerFormat: "JWT",
        description: "JWT Bearer token authentication",
      },
    },
  };

  const fullCard: AgentCard = {
    name: "code-review-agent",
    description:
      "An autonomous code review agent that analyzes pull requests and provides actionable feedback.",
    version: "1.2.0",
    supportedInterfaces: [],
    capabilities: {
      streaming: true,
      pushNotifications: true,
      extendedAgentCard: false,
    },
    skills: [
      {
        id: "review-pr",
        name: "Pull Request Review",
        description:
          "Analyzes code diffs and produces structured review comments.",
        tags: ["code-review", "static-analysis", "security"],
        examples: [
          "Review this PR for security issues",
          "Check for performance regressions in the diff",
        ],
      },
      {
        id: "suggest-fix",
        name: "Suggest Fix",
        description: "Generates code fix suggestions for identified issues.",
        tags: ["code-generation", "refactoring"],
        inputModes: ["text/plain"],
        outputModes: ["text/plain", "application/json"],
      },
    ],
    defaultInputModes: ["text/plain", "application/json"],
    defaultOutputModes: ["text/plain", "application/json"],
    provider: {
      url: "https://brainwires.dev",
      organization: "Brainwires",
    },
    securitySchemes,
    documentationUrl: "https://docs.brainwires.dev/agents/code-review",
    iconUrl: "https://brainwires.dev/icons/code-review.svg",
  };

  // -----------------------------------------------------------------------
  // 2. Serialize to JSON
  // -----------------------------------------------------------------------
  const json = JSON.stringify(fullCard, null, 2);
  console.log(json);
  console.log();

  // -----------------------------------------------------------------------
  // 3. Round-trip: deserialize and verify
  // -----------------------------------------------------------------------
  console.log("=== Round-Trip Verification ===\n");

  const deserialized: AgentCard = JSON.parse(json);

  console.assert(
    deserialized.name === fullCard.name,
    "round-trip name mismatch",
  );
  console.assert(
    deserialized.skills.length === fullCard.skills.length,
    "round-trip skills count mismatch",
  );
  console.log(
    `Round-trip OK: name = "${deserialized.name}", skills = ${deserialized.skills.length}`,
  );
  console.log();

  // -----------------------------------------------------------------------
  // 4. Build a minimal AgentCard (bare essentials only)
  // -----------------------------------------------------------------------
  console.log("=== Minimal AgentCard ===\n");

  const minimalCard: AgentCard = {
    name: "echo-agent",
    description: "A minimal agent that echoes messages back.",
    version: "0.8.0",
    supportedInterfaces: [],
    capabilities: {},
    skills: [],
    defaultInputModes: ["text/plain"],
    defaultOutputModes: ["text/plain"],
  };

  const minimalJson = JSON.stringify(minimalCard, null, 2);
  console.log(minimalJson);
  console.log();

  // -----------------------------------------------------------------------
  // 5. Side-by-side comparison
  // -----------------------------------------------------------------------
  console.log("=== Comparison ===\n");

  const pad = (s: string, n: number) => s.padEnd(n);

  console.log(
    `${pad("Field", 20)} ${pad("Full Card", 25)} ${pad("Minimal Card", 25)}`,
  );
  console.log("-".repeat(70));
  console.log(
    `${pad("name", 20)} ${pad(fullCard.name, 25)} ${pad(minimalCard.name, 25)}`,
  );
  console.log(
    `${pad("version", 20)} ${pad(fullCard.version, 25)} ${
      pad(minimalCard.version, 25)
    }`,
  );
  console.log(
    `${pad("skills", 20)} ${pad(String(fullCard.skills.length), 25)} ${
      pad(String(minimalCard.skills.length), 25)
    }`,
  );
  console.log(
    `${pad("streaming", 20)} ${
      pad(String(fullCard.capabilities.streaming), 25)
    } ${pad(String(minimalCard.capabilities.streaming), 25)}`,
  );
  console.log(
    `${pad("provider", 20)} ${
      pad(fullCard.provider?.organization ?? "None", 25)
    } ${pad(minimalCard.provider?.organization ?? "None", 25)}`,
  );
  console.log(
    `${pad("securitySchemes", 20)} ${
      pad(String(fullCard.securitySchemes !== undefined), 25)
    } ${pad(String(minimalCard.securitySchemes !== undefined), 25)}`,
  );

  console.log("\nDone.");
}

await main();

File: crates/rullama-a2a/examples/agent_card.rs
//! Agent Card Construction Example
//!
//! Demonstrates building A2A AgentCards with capabilities, skills,
//! security schemes, and round-trip JSON serialization.

use rullama_a2a::{
    AgentCapabilities, AgentCard, AgentProvider, AgentSkill, HttpAuthSecurityScheme, SecurityScheme,
};
use std::collections::HashMap;

fn main() {
    // 1. Build a full AgentCard with all fields populated
    println!("=== Full AgentCard ===\n");

    let mut security_schemes = HashMap::new();
    security_schemes.insert(
        "bearer_auth".to_string(),
        SecurityScheme {
            api_key: None,
            http_auth: Some(HttpAuthSecurityScheme {
                scheme: "Bearer".to_string(),
                bearer_format: Some("JWT".to_string()),
                description: Some("JWT Bearer token authentication".to_string()),
            }),
            oauth2: None,
            open_id_connect: None,
            mtls: None,
        },
    );

    let full_card = AgentCard {
        name: "code-review-agent".to_string(),
        description: "An autonomous code review agent that analyzes pull requests \
                      and provides actionable feedback."
            .to_string(),
        version: "1.2.0".to_string(),
        supported_interfaces: vec![],
        capabilities: AgentCapabilities {
            streaming: Some(true),
            push_notifications: Some(true),
            extended_agent_card: Some(false),
            extensions: None,
        },
        skills: vec![
            AgentSkill {
                id: "review-pr".to_string(),
                name: "Pull Request Review".to_string(),
                description: "Analyzes code diffs and produces structured review comments."
                    .to_string(),
                tags: vec![
                    "code-review".to_string(),
                    "static-analysis".to_string(),
                    "security".to_string(),
                ],
                examples: Some(vec![
                    "Review this PR for security issues".to_string(),
                    "Check for performance regressions in the diff".to_string(),
                ]),
                input_modes: None,
                output_modes: None,
                security_requirements: None,
            },
            AgentSkill {
                id: "suggest-fix".to_string(),
                name: "Suggest Fix".to_string(),
                description: "Generates code fix suggestions for identified issues.".to_string(),
                tags: vec!["code-generation".to_string(), "refactoring".to_string()],
                examples: None,
                input_modes: Some(vec!["text/plain".to_string()]),
                output_modes: Some(vec![
                    "text/plain".to_string(),
                    "application/json".to_string(),
                ]),
                security_requirements: None,
            },
        ],
        default_input_modes: vec!["text/plain".to_string(), "application/json".to_string()],
        default_output_modes: vec!["text/plain".to_string(), "application/json".to_string()],
        provider: Some(AgentProvider {
            url: "https://brainwires.dev".to_string(),
            organization: "Brainwires".to_string(),
        }),
        security_schemes: Some(security_schemes),
        security_requirements: None,
        documentation_url: Some("https://docs.brainwires.dev/agents/code-review".to_string()),
        icon_url: Some("https://brainwires.dev/icons/code-review.svg".to_string()),
        signatures: None,
    };

    // 2. Serialize to JSON
    let json = serde_json::to_string_pretty(&full_card).expect("serialize full card");
    println!("{json}\n");

    // 3. Round-trip: deserialize and verify
    println!("=== Round-Trip Verification ===\n");

    let deserialized: AgentCard = serde_json::from_str(&json).expect("deserialize full card");
    assert_eq!(
        deserialized.name, full_card.name,
        "round-trip name mismatch"
    );
    assert_eq!(
        deserialized.skills.len(),
        full_card.skills.len(),
        "round-trip skills count mismatch"
    );
    println!(
        "Round-trip OK: name = {:?}, skills = {}\n",
        deserialized.name,
        deserialized.skills.len()
    );

    // 4. Build a minimal AgentCard (bare essentials only)
    println!("=== Minimal AgentCard ===\n");

    let minimal_card = AgentCard {
        name: "echo-agent".to_string(),
        description: "A minimal agent that echoes messages back.".to_string(),
        version: "0.7.0".to_string(),
        supported_interfaces: vec![],
        capabilities: AgentCapabilities::default(),
        skills: vec![],
        default_input_modes: vec!["text/plain".to_string()],
        default_output_modes: vec!["text/plain".to_string()],
        provider: None,
        security_schemes: None,
        security_requirements: None,
        documentation_url: None,
        icon_url: None,
        signatures: None,
    };

    let minimal_json = serde_json::to_string_pretty(&minimal_card).expect("serialize minimal card");
    println!("{minimal_json}\n");

    // 5. Side-by-side summary
    println!("=== Comparison ===\n");
    println!("{:<20} {:<25} {:<25}", "Field", "Full Card", "Minimal Card");
    println!("{:-<70}", "");
    println!(
        "{:<20} {:<25} {:<25}",
        "name", full_card.name, minimal_card.name
    );
    println!(
        "{:<20} {:<25} {:<25}",
        "version", full_card.version, minimal_card.version
    );
    println!(
        "{:<20} {:<25} {:<25}",
        "skills",
        full_card.skills.len(),
        minimal_card.skills.len()
    );
    println!(
        "{:<20} {:<25} {:<25}",
        "streaming",
        format!("{:?}", full_card.capabilities.streaming),
        format!("{:?}", minimal_card.capabilities.streaming)
    );
    println!(
        "{:<20} {:<25} {:<25}",
        "provider",
        full_card
            .provider
            .as_ref()
            .map(|p| p.organization.as_str())
            .unwrap_or("None"),
        minimal_card
            .provider
            .as_ref()
            .map(|p| p.organization.as_str())
            .unwrap_or("None")
    );
    println!(
        "{:<20} {:<25} {:<25}",
        "security_schemes",
        full_card.security_schemes.is_some(),
        minimal_card.security_schemes.is_some()
    );
}
